jcapi-tags.go

```diff
@@ -42,84 +51,59 @@ func GetTagNames(tags []JCTag) []string {
 	return returnVal
 }
 
-func (tag *JCTag) MarshalJSON() ([]byte, error) {
+func getJCTagsFromResults(result []byte) (tags []JCTag, err JCError) {
+	tag := JCTag{}
 
-	var builder []string
+	// Try unmarshalling as a single tag first, and if that works, stop there
+	err = json.Unmarshal(result, &tag)
+	if err != nil || tag.Id == "" {
+		err = nil
 
-	if tag.Id != "" {
-		builder = append(builder, buildJSONKeyValuePair("_id", tag.Id))
-	}
+		tagResults := JCTagResults{}
 
-	builder = append(builder, buildJSONKeyValuePair("name", tag.Name))
-	builder = append(builder, buildJSONKeyValuePair("groupname", tag.GroupName))
-	builder = append(builder, buildJSONStringArray("systems", tag.Systems))
-	builder = append(builder, buildJSONStringArray("systemusers", tag.SystemUsers))
-	builder = append(builder, buildJSONStringArray("regularExpressions", tag.RegularExpressions))
-	builder = append(builder, buildJSONKeyValuePair("expirationTime", tag.ExpirationTime))
-	builder = append(builder, buildJSONKeyValueBoolPair("expired", tag.Expired))
-	builder = append(builder, buildJSONKeyValueBoolPair("selected", tag.Selected))
+		// Nope, must be a results array...
+		err = json.Unmarshal(result, &tagResults)
+		if err != nil {
+			err = fmt.Errorf("Could not unmarshal result '%s', err='%s'", string(result), err.Error())
+		}
 
-	builder = append(builder, buildJSONKeyValueBoolPair("externallyManaged", tag.ExternallyManaged))
-	builder = append(builder, buildJSONKeyValuePair("externalDN", tag.ExternalDN))
-	builder = append(builder, buildJSONKeyValuePair("externalSourceType", tag.ExternalSourceType))
+		tags = tagResults.Results
+	} else {
+		tags = append(tags, tag)
```

Comment:
The append here, while correct, I think implies that `tags` might have other data, which it can't here.  I think I'd write this as `tags = []JCTag{tag}` so that its clear in one part of the if block you set it to `.Results` and in the other, a single item array.